frontend/packages/dev-console/src/components/pipelineruns/logs/Logs.tsx

```diff
@@ -0,0 +1,133 @@
+import { coFetchText } from '@console/internal/co-fetch';
+import './Logs.scss';
+
+type LogsProps = {
+  resource: PodKind;
+  resourceStatus: string;
+  container: ContainerSpec;
+  render: boolean;
+  autoScroll?: boolean;
+  onComplete: (containerName: string) => void;
+};
+
+const Logs: React.FC<LogsProps> = ({
+  resource,
+  resourceStatus,
+  container,
+  onComplete,
+  render,
+  autoScroll = true,
+}) => {
+  const { name } = container;
+  const { kind, metadata = {} } = resource;
+  const { name: resName, namespace: resNamespace } = metadata;
+  const contentRef = React.useRef<HTMLDivElement>(null);
+  const [error, setError] = React.useState<boolean>(false);
+  const resourceStatusRef = React.useRef<string>(resourceStatus);
+  const onCompleteRef = React.useRef<(name) => void>();
+  onCompleteRef.current = onComplete;
+  const appendMessage = React.useRef<(blockContent) => void>();
+  appendMessage.current = React.useCallback(
+    (blockContent) => {
+      const contentLines = blockContent.split('\n').filter((line) => !!line);
+      if (contentRef.current && contentLines.length >= 0) {
+        const elements = contentLines.map((content) => {
+          const customElement = document.createElement('div');
+          customElement.textContent = content;
+          return customElement;
+        });
+        elements.forEach((element) => {
+          contentRef.current.append(element);
+        });
+        const lastElement = elements[elements.length - 1];
+        if (render && lastElement && autoScroll) {
+          lastElement.scrollIntoView({ behavior: 'smooth', block: 'end' });
+        }
+      }
+    },
+    [autoScroll, render],
+  );
+
+  if (resourceStatusRef.current !== resourceStatus) {
+    resourceStatusRef.current = resourceStatus;
+  }
+
+  React.useEffect(() => {
+    let loaded: boolean = false;
+    let ws: WSFactory;
+    const urlOpts = {
+      ns: resNamespace,
+      name: resName,
+      path: 'log',
+      queryParams: {
+        container: name,
+        follow: 'true',
+      },
+    };
+    const watchURL = resourceURL(modelFor(kind), urlOpts);
+    if (resourceStatusRef.current === LOG_SOURCE_TERMINATED) {
+      coFetchText(watchURL)
+        .then((res) => {
+          if (loaded) return;
+          appendMessage.current(res);
+          onCompleteRef.current(name);
+        })
+        .catch(() => {
+          if (loaded) return;
+          setError(true);
+          onCompleteRef.current(name);
+        });
+    } else {
+      const wsOpts = {
+        host: 'auto',
+        path: watchURL,
+        subprotocols: ['base64.binary.k8s.io'],
+      };
+      ws = new WSFactory(watchURL, wsOpts);
+      ws.onmessage((msg) => {
+        if (loaded) return;
+        const message = Base64.decode(msg);
+        appendMessage.current(message);
+      })
+        .onclose(() => {
+          onCompleteRef.current(name);
+        })
+        .onerror(() => {
```

Comment:
Something is fishy here... I tried to find the cause but I'm a bit perplexed. On the first step of each task, it errors out with a 400 (client) error. The websocket logs are entirely unhelpful and the error is the websocket is closed ([readyState](https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/readyState) === 3).

![image](https://user-images.githubusercontent.com/8126518/78929289-1a09d800-7a70-11ea-9b0a-f60ccd303794.png)

Pods do not have this issue 😕 If I go to the Pod fast enough, I see it log out without an issue and then close the websocket. Not sure what we are doing differently here. 

We should definitely track this down, but I think we can leave this for now and swing back in 4.5 bug fixing. This is super odd and we are clearly missing something here. Please log a bug to track this @sahil143 

Also... Console does a `retry` shindig if they showed an error:
```
          <Alert
            isInline
            className="co-alert"
            variant="danger"
            title="An error occurred while retrieving the requested logs."
            action={<AlertActionLink onClick={this._restartStream}>Retry</AlertActionLink>}
          />
```

Reply:
@andrewballantyne I only got this when Websocket request timed out with an error
```
WebSocket connection to 'ws://localhost:9000/api/kubernetes/api/v1/namespaces/t0s/pods/nodejs-ex-git-a01t7k-build-7d64p-pod-6t9j2/log?container=step-create-dir-image-p8tlw&follow=true' failed: Error during WebSocket handshake: Unexpected response code: 400
```
Otherwise I'm getting proper logs
![Screenshot from 2020-04-10 00-44-32](https://user-images.githubusercontent.com/9278015/78932355-f329c100-7ac4-11ea-96f9-9ff2527a8275.png)

Reply:
Right, and it happened for me every time (tested it at least half a dozen times in a row) for that first task while it was being fetched by the websocket. If you come to this log/task after that first one is done, I see no error.

Still don't see this on the Pod log, clearly something is different. Since it's not critical and we have some bandwidth next sprint, let's just log something to investigate.

Reply:
Raised a Jira ticket to track this https://issues.redhat.com/browse/ODC-3540.